exp/api/v1beta1/azuremanagedmachinepool_webhook.go

```diff
@@ -319,3 +328,50 @@ func (m *AzureManagedMachinePool) validateEnableNodePublicIP() error {
 	}
 	return nil
 }
+
+// validateKubeletConfig enforces the AKS API configuration for KubeletConfig.
+// See:  https://learn.microsoft.com/en-us/azure/aks/custom-node-configuration.
+func (m *AzureManagedMachinePool) validateKubeletConfig() error {
+	var allowedUnsafeSysctlsPatterns = []string{
+		`^kernel\.shm.+$`,
+		`^kernel\.msg.+$`,
+		`^kernel\.sem$`,
+		`^fs\.mqueue\..+$`,
+		`^net\..+$`,
+	}
+	if m.Spec.KubeletConfig != nil {
+		if m.Spec.KubeletConfig.CPUCfsQuotaPeriod != nil {
+			if !strings.HasSuffix(to.String(m.Spec.KubeletConfig.CPUCfsQuotaPeriod), "ms") {
+				return field.Invalid(
+					field.NewPath("Spec", "KubeletConfig", "CPUCfsQuotaPeriod"),
+					m.Spec.KubeletConfig.CPUCfsQuotaPeriod,
+					"must be a string value in milliseconds with a 'ms' suffix, e.g., '100ms'")
+			}
+		}
+		if m.Spec.KubeletConfig.ImageGcHighThreshold != nil && m.Spec.KubeletConfig.ImageGcLowThreshold != nil {
+			if to.Int32(m.Spec.KubeletConfig.ImageGcLowThreshold) > to.Int32(m.Spec.KubeletConfig.ImageGcHighThreshold) {
+				return field.Invalid(
+					field.NewPath("Spec", "KubeletConfig", "ImageGcLowThreshold"),
+					m.Spec.KubeletConfig.ImageGcLowThreshold,
+					fmt.Sprintf("must not be greater than ImageGcHighThreshold, ImageGcLowThreshold=%d, ImageGcHighThreshold=%d",
+						to.Int32(m.Spec.KubeletConfig.ImageGcLowThreshold), to.Int32(m.Spec.KubeletConfig.ImageGcHighThreshold)))
+			}
+		}
+		for _, val := range m.Spec.KubeletConfig.AllowedUnsafeSysctls {
+			var hasMatch bool
+			for _, p := range allowedUnsafeSysctlsPatterns {
+				if m, _ := regexp.MatchString(p, val); m {
+					hasMatch = true
+					break
+				}
+			}
+			if !hasMatch {
+				return field.Invalid(
+					field.NewPath("Spec", "KubeletConfig", "AllowedUnsafeSysctls"),
+					m.Spec.KubeletConfig.AllowedUnsafeSysctls,
+					fmt.Sprintf("%s is not a supported AllowedUnsafeSysctls configuration", val))
+			}
+		}
```

Comment:
I was looking at two for-loops and thought if I we could make it single loop somehow. Came up with this implementation https://go.dev/play/p/p2qk7dJtOyX. What do you think? 

I also noticed that a similar implementation as that of https://go.dev/play/p/p2qk7dJtOyX will give us a much granular check. We can point to each item in `m.Spec.KubeletConfig.AllowedUnsafeSysctls` that don't match the allowed values. I am not sure if it would be over-engineering or needed here, but would like your thoughts on it too.

Reply:
The downside to those approaches is that we have to compile the regex patterns in each webhook flow, whether or not there are any `AllowedUnsafeSysctls` in the array. wdyt?

Reply:
Tl;DR Makes sense, we can ignore optimization here for this use case. 

> we have to compile the regex patterns in each webhook flow

Thank you for reminding this, I agree.

>  whether or not there are any AllowedUnsafeSysctls in the array

We could possibly compile the regular expressions after checking the size of `AllowedUnsafeSysctls`. But would be an overkill for our scenario. 
Besides I like the current implementation where we return error on the first occurrence of error rather than making a list of all errors. Blocking DoS attempts.

Reply:
Thanks again for this thread, optimization explorations like this that run regularly in the background are very welcome!